api/wft-complete-metadata.md

```diff
@@ -0,0 +1,133 @@
+# Workflow Task Completion Metadata
+
+We have a need to attach various bits of metadata to workflow task completions. There are a number
+of usecases in which the SDK would like to save some information for itself, or tell server about
+something, but this information does not influence replay or the reconstitution of workfow state in
+any way. Thus, it shouldn't go in a marker or some new kind of history event, because this would
+violate the (useful) idea that history events are primarily concerned with recreating workflow
+state.
+
+## Use cases
+
+### Reporting currently-being-awaited commands & facts about them
+
+Right now, besides a query, there's no simple way for a user to understand what their workflow is
+actively waiting on. Solving this would make it much easier for users to have insight into why
+their workflow is or isn't making progress, and could be nicely exposed in the UI.
+
+In particular, understanding what's going on while your workflow is currently running local
+activities and performing WFT heartbeats is very opaque. This would make that much more clear.
+
+Imagine being able to get information like this (and do useful things with it in the UI):
+
+```
+waiting_on: [
```

Comment:
I don't think we want to add this type of data.
We already know which activities and timers are unresolved from workflow history.
We don't know what the workflow code is effectively blocked on and we can't know that.

Reply:
Same. But I consider almost all of this as "could use metadata for one day" and not necessarily now.

Reply:
How would we not want to add this? This is incredibly useful for a user trying to understand what their workflow is currently waiting on.

Local activities are the prime example, right there, and they are *not* already in history. What rationale is there for not adding this? 

It's also not true that we don't know what code is blocked on. We know exactly what it's blocked on, because that's the definition of when you finish an activation / workflow task - when all routines are at an await point-type thing. They don't all necessarily correspond 1:1 to commands, but you know exactly what you're blocked on, by definition.

Reply:
What is this code blocked on?

```ts
export async function myWorkflow() {
  // run in the background and ignore if workflow is completed
  someActivity();
  await CancellationScope.current().cancelRequested;
}
```

You can tell by the stack trace (especially the structured one where users can see the actual code) but the SDK would struggle knowing which promises are "blocking" the workflow from making progress.

Local activities are definitely more interesting here since they don't show up in history until completed but even then I think we can infer from the fact that a WFT heartbeat is present that local activities are running and instruct the user to go to the stack trace query to get more information.

Reply:
It's blocked on waiting for the current cancellation scope to be requested. It's right there, where you're awaiting.

Admittedly that's harder to determine in TS than some other languages, but it's still 100% possible and useful. In Rust it's trivial. In Python it's probably pretty easy too.

Why would I want to _infer_ that we're probably waiting on some local activities when I could know exactly how many are currently running and their types? I don't really follow what the argument is for _not_ doing it? Besides that, we need this for cloud, which is one of the motivating reasons.

Reply:
My point is that we already know this from the stack trace, no need to add this data to histories.

Reply:
Stack trace is a point-in-time view. This is a progressive view that works for every WFT, not just wherever you currently are.

A good point is that "currently waiting on" and "currently outstanding" are worth distinguishing. We need to know currently outstanding LAs for cloud, and perhaps things that are being waiting on is fine to leave for stack traces, but seems to me this would still be remarkably useful for debugging the progress of a workflow and it's not going to take up much space in history.